### ISSUE TITLE
[Bug?]: Yarn 3+ does not connect properly some dependencies, making them unreachable

### ISSUE THREAD

**Issue by dubzzz** [yarnpkg/berry#5082]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

My package `pure-rand` is using a library called `fast-check` to test itself (declared as a devDependency). Unfortunately when installing this dependency locally, yarn somehow skip some of the dependencies needed by `fast-check` to be able to run.

More precisely, under the hood, `fast-check` (test framework) is leveraging `pure-rand`, so Yarn somehow skips it: neither installed from npm nor simlink-ed. Whenever entering a file asking the require/import of `pure-rand`, node fails with:

```txt
Error: Cannot find module 'pure-rand'
```

### To reproduce

I tried to extract a minimal repro of the issue at: https://github.com/dubzzz/yarn-3-repro-pure-rand

Basically here are the two main characteristics that make it fails:
- package name matches the one asked by `fast-check`
- version matches the requirements of `fast-check`

On my side, I was either expecting Yarn to pull the version of pure-rand from the npm registry or to simlink the one defined within the repository.

### Environment

```shell
System:
    OS: Windows 10 10.0.22623
    CPU: (8) x64 Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
  Binaries:
    Node: 16.14.0 - C:\Users\N4442~1.DUB\AppData\Local\Temp\xfs-67e9fc28\node.CMD
    Yarn: 3.3.0 - C:\Users\N4442~1.DUB\AppData\Local\Temp\xfs-67e9fc28\yarn.CMD
    npm: 8.3.1 - C:\Program Files\nodejs\npm.CMD
```

### Additional context

If the "bug" is confirmed and depending how complex it is to fix, with good pointers I could potentially give a hand and fix it.

**Comment by arcanis** [comment]
Indeed, I don't see the nm self-reference symlink to `pure-rand`, even on latest RCs 🤔

Does that raise a bell, @larixer ?

**Comment by larixer** [comment]
And why do you call the root package as `pure-rand`, obviously the Yarn is confused by it, it thinks that the dependency is met,  because root package is called `pure-rand`. Call it `pure-rand-foo` and everything will work.

**Comment by dubzzz** [comment]
I call it `pure-rand` because I'm the author of `pure-rand` and try to move to Yarn 3+ at the moment. The linked repository is just a toy reproduction of the issue I have on https://github.com/dubzzz/pure-rand/pull/465. But so far I cannot use Yarn 3+ for `pure-rand`, or at least don't know how to do so.

_And tomorrow, I may have the same issue with `fast-check` as it might be pulled by `jest` soon or later and be part of its dependencies._

**Comment by larixer** [comment]
Yes, why do you call it the same name as one of your dependencies though, it is the reason for your problem, call it something else

**Comment by larixer** [comment]
@dubzzz Thanks, I will reevalute tomorrow thoroughly with @arcanis, looks like he understand your issue and I will come back to you

**Comment by dubzzz** [comment]
No worry, no urgency at all.

**Comment by larixer** [comment]
> Indeed, I don't see the nm self-reference symlink to pure-rand, even on latest RCs 🤔
>
> Does that raise a bell, @larixer ?

@arcanis Looks like this is a problem in the core. I have tried to use `pnp` linker with this project and I see two problems:
1. There is no `pure-rand` dependency from `npm` in the cache directory.
2. The state in `pnp` file is incorrect as well:
```json
    ["fast-check", [\
      ["npm:3.3.0", {\
        "packageLocation": "../../../.yarn/berry/cache/fast-check-npm-3.3.0-4a9a3ca773-9.zip/node_modules
        "packageDependencies": [\
          ["fast-check", "npm:3.3.0"],\
          ["pure-rand", "workspace:."]\
          ^^^ should be something like "pure-rand": "npm:5.0.2" instead here
        ],\
        "linkType": "HARD"\
      }]\
    ]],\
```

**Comment by larixer** [comment]
@dubzzz Looks like this is expected behaviour. You can change it via `enableTransparentWorkspaces: false`, in this case the `pure-rand` dependency will be downloaded from `npm registry`.

**Comment by arcanis** [comment]
To be clear, I think the node-modules linker should still create a symlink in `node_modules/pure-rand/node_modules/fast-check -> ../../../`, but yep, `enableTransparentWorkspaces: false` should do the trick if you're fine with using the your own package as any other dependency (it's what Babel does as well, since they depend on themselves to compile their own sources).

**Comment by larixer** [comment]
> To be clear, I think the node-modules linker should still create a symlink in node_modules/pure-rand/node_modules/fast-check -> ../../../, but yep, enableTransparentWorkspaces: false should do the trick if you're fine with using the your own package as any other dependency (it's what Babel does as well, since they depend on themselves to compile their own sources).

I strongly disagree. This will create symbolic link loop, which is poorly supported by dependencies and tools as reported for example by:
https://github.com/yarnpkg/berry/issues/4232
https://github.com/yarnpkg/berry/issues/3256
https://github.com/yarnpkg/berry/issues/3409
https://github.com/yarnpkg/berry/issues/4232

**Comment by larixer** [comment]
> Yes, that's true - but we already have nmSelfReferences to disable the symlink loop behaviour, so we could create or not the symlink based on that.

I agree with the intent, but the implementation should not involve symlink loops anyway, I strongly against symlink loops. The implementation can use symlink to intermediate storage folder to avoid symlink and providing self-reference at the same time. This also could be used not only for self-references but also to better support aliased packages that provide themselves as peer dependencies.

**Comment by dubzzz** [comment]
Thanks a lot `enableTransparentWorkspaces` is definitely what I was looking for. For links within the workspace itself I'll definitely use the `workspace:*` notation.
